File: madhhab_service.py
import logging
import logging
from typing import Dict, List, Optional
from madhahib import get_madhhab, get_all_madhahib, MADHAHIB

logger = logging.getLogger(__name__)

class MadhahibService:
    """Service class to handle all madhhab-related operations"""
    
    def __init__(self):
        self.madhahib = MADHAHIB
    
    def get_available_madhahib(self) -> List[str]:
        """Get list of available madhhab names"""
        return list(self.madhahib.keys())
    
    def get_madhhab_info(self, madhhab_name: str) -> Optional[Dict]:
        """Get basic information about a specific madhhab"""
        madhhab = get_madhhab(madhhab_name)
        if madhhab:
            return madhhab.get_basic_info()
        return None
    
    def get_single_madhhab_response(self, madhhab_name: str, question: str, api_key: str) -> str:
        """Get response from a single madhhab"""
        try:
            madhhab = get_madhhab(madhhab_name)
            if not madhhab:
                return f"المذهب {madhhab_name} غير متوفر"
            
            # Use sync method since Gemini API doesn't support async
            import google.generativeai as genai
            genai.configure(api_key=api_key)
            model = genai.GenerativeModel('models/gemini-2.0-flash-001')
            
            prompt = madhhab.build_prompt(question)
            response = model.generate_content(prompt)
            
            if response and response.text:
                logger.info(f"Successfully received response from {madhhab_name}")
                return response.text
            else:
                logger.warning(f"Empty response from {madhhab_name}")
                return f"عذراً، لم أتمكن من الحصول على إجابة من المذهب {madhhab_name}"
            
        except Exception as e:
            logger.error(f"Error getting response from {madhhab_name}: {str(e)}")
            return f"حدث خطأ أثناء الحصول على إجابة من المذهب {madhhab_name}: {str(e)}"
    
    def get_all_madhahib_responses(self, question: str, api_key: str) -> Dict[str, str]:
        """Get responses from all madhahib sequentially"""
        try:
            all_madhahib = get_all_madhahib()
            results = {}
            
            import google.generativeai as genai
            genai.configure(api_key=api_key)
            model = genai.GenerativeModel('models/gemini-2.0-flash-001')
            
            for madhhab_name, madhhab_instance in all_madhahib.items():
                try:
                    logger.info(f"Getting response from {madhhab_name}...")
                    prompt = madhhab_instance.build_prompt(question)
                    response = model.generate_content(prompt)
                    
                    if response and response.text:
                        results[madhhab_name] = response.text
                        logger.info(f"Successfully received response from {madhhab_name}")
                    else:
                        results[madhhab_name] = f"عذراً، لم أتمكن من الحصول على إجابة من المذهب {madhhab_name}"
                        logger.warning(f"Empty response from {madhhab_name}")
                        
                except Exception as e:
                    logger.error(f"Error getting response from {madhhab_name}: {str(e)}")
                    results[madhhab_name] = f"حدث خطأ أثناء الحصول على إجابة من المذهب {madhhab_name}"
            
            return results
            
        except Exception as e:
            logger.error(f"Error getting responses from all madhahib: {str(e)}")
            return {"error": f"حدث خطأ أثناء الحصول على الإجابات: {str(e)}"}
    
    def format_combined_response(self, responses: Dict[str, str]) -> str:
        """Format multiple madhhab responses into a single combined response"""
        combined_answer = ""
        
        for madhhab_name, answer in responses.items():
            if madhhab_name != "error":
                combined_answer += f"\n\n{'='*50}\n"
                combined_answer += f"إجابة المذهب {madhhab_name}\n"
                combined_answer += f"{'='*50}\n\n"
                combined_answer += answer
        
        if "error" in responses:
            combined_answer += f"\n\n⚠️ تنبيه: {responses['error']}"
        
        return combined_answer
    
    def get_madhhab_comparison(self, madhhab_names: List[str]) -> Dict[str, Dict]:
        """Compare specific madhahib and return their key differences"""
        comparison = {}
        
        for madhhab_name in madhhab_names:
            madhhab = get_madhhab(madhhab_name)
            if madhhab:
                comparison[madhhab_name] = {
                    'methodology': madhhab.get_methodology(),
                    'main_sources': madhhab.main_sources,
                    'famous_scholars': madhhab.famous_scholars[:3],  # Top 3 scholars
                    'geographic_influence': madhhab.geographic_influence[:3]  # Top 3 regions
                }
        
        return comparison

File: src/services/madhhabService.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { config } from '@/config';
import { MadhhabInfo, MadhhabResponse, ApiResponse } from '@/types';

export class MadhhabService {
  private genAI: GoogleGenerativeAI;
  private model: any;

  constructor(apiKey?: string) {
    const key = apiKey || config.api.gemini.defaultKey;
    this.genAI = new GoogleGenerativeAI(key);
    this.model = this.genAI.getGenerativeModel({ model: config.api.gemini.model });
  }

  // الحصول على معلومات المذاهب
  getMadhahibInfo(): Record<string, MadhhabInfo> {
    return config.madhahib;
  }

  // الحصول على معلومات مذهب واحد
  getMadhhabInfo(madhhabName: string): MadhhabInfo | null {
    const madhahib = this.getMadhahibInfo();
    return madhahib[madhhabName] || null;
  }

  // بناء التوجيه لمذهب معين
  private buildMadhhabPrompt(madhhabInfo: MadhhabInfo, question: string): string {
    return `أنا عالم إسلامي متخصص في المذهب ${madhhabInfo.name}. 
    تحدث باسم المذهب ${madhhabInfo.name} وأجب على السؤال التالي من منظور هذا المذهب:
    
    السؤال: ${question}
    
    ابدأ إجابتك بـ: "أنا المذهب ${madhhabInfo.name}..."
    وتحدث بصيغة المتكلم كأنك المذهب نفسه.
    
    يجب أن تتضمن الإجابة:
    1. تعريف بالمذهب ومؤسسه (${madhhabInfo.founder})
    2. الحكم الشرعي حسب هذا المذهب
    3. الدليل من القرآن أو السنة
    4. أقوال علماء هذا المذهب
    5. المنهجية المميزة لهذا المذهب في الاستنباط
    6. التعليل والحكمة
    
    معلومات إضافية عن المذهب ${madhhabInfo.name}:
    - أشتهر بـ: ${madhhabInfo.methodology}
    - المصادر المعتمدة: ${madhhabInfo.sources.join('، ')}
    - من أبرز علمائه: ${madhhabInfo.scholars.slice(0, 3).join('، ')}
    - ينتشر في: ${madhhabInfo.influence.slice(0, 3).join('، ')}
    
    تأكد من إبراز هذه الخصائص في إجابتك وكيف تؤثر على الحكم الفقهي.`;
  }

  // الحصول على إجابة من مذهب واحد
  async getSingleMadhhabResponse(
    madhhabName: string,
    question: string
  ): Promise<ApiResponse<MadhhabResponse>> {
    try {
      const madhhabInfo = this.getMadhhabInfo(madhhabName);
      if (!madhhabInfo) {
        return {
          status: 'error',
          message: `المذهب ${madhhabName} غير متوفر`,
        };
      }

      const prompt = this.buildMadhhabPrompt(madhhabInfo, question);
      const result = await this.model.generateContent(prompt);
      const response = await result.response;
      const text = response.text();

      if (!text) {
        return {
          status: 'error',
          message: `عذراً، لم أتمكن من الحصول على إجابة من المذهب ${madhhabInfo.name}`,
        };
      }

      return {
        status: 'success',
        data: {
          madhhab: madhhabInfo.name,
          answer: text,
          timestamp: new Date().toISOString(),
        },
      };
    } catch (error) {
      console.error(`Error getting response from ${madhhabName}:`, error);
      return {
        status: 'error',
        message: `حدث خطأ أثناء الحصول على إجابة من المذهب ${madhhabName}`,
      };
    }
  }

  // الحصول على إجابات من جميع المذاهب
  async getAllMadhahibResponses(
    question: string
  ): Promise<ApiResponse<Record<string, MadhhabResponse>>> {
    try {
      const madhahib = this.getMadhahibInfo();
      const responses: Record<string, MadhhabResponse> = {};

      for (const [key, madhhabInfo] of Object.entries(madhahib)) {
        try {
          const prompt = this.buildMadhhabPrompt(madhhabInfo, question);
          const result = await this.model.generateContent(prompt);
          const response = await result.response;
          const text = response.text();

          if (text) {
            responses[madhhabInfo.name] = {
              madhhab: madhhabInfo.name,
              answer: text,
              timestamp: new Date().toISOString(),
            };
          } else {
            responses[madhhabInfo.name] = {
              madhhab: madhhabInfo.name,
              answer: `عذراً، لم أتمكن من الحصول على إجابة من المذهب ${madhhabInfo.name}`,
              timestamp: new Date().toISOString(),
            };
          }
        } catch (error) {
          console.error(`Error getting response from ${madhhabInfo.name}:`, error);
          responses[madhhabInfo.name] = {
            madhhab: madhhabInfo.name,
            answer: `حدث خطأ أثناء الحصول على إجابة من المذهب ${madhhabInfo.name}`,
            timestamp: new Date().toISOString(),
          };
        }
      }

      return {
        status: 'success',
        data: responses,
      };
    } catch (error) {
      console.error('Error getting responses from all madhahib:', error);
      return {
        status: 'error',
        message: 'حدث خطأ أثناء الحصول على الإجابات',
      };
    }
  }

  // تنسيق الإجابة المجمعة
  formatCombinedResponse(responses: Record<string, MadhhabResponse>): string {
    let combinedAnswer = '';
    
    for (const [madhhabName, response] of Object.entries(responses)) {
      combinedAnswer += `\n\n${'='.repeat(50)}\n`;
      combinedAnswer += `إجابة المذهب ${madhhabName}\n`;
      combinedAnswer += `${'='.repeat(50)}\n\n`;
      combinedAnswer += response.answer;
    }
    
    return combinedAnswer;
  }
}
